test/clj/game/cards/ice_test.clj

```diff
@@ -1685,37 +1685,116 @@
+                                   "Project Vitruvius"
+                                   "NEXT Wave 2"
+                                   "Accelerated Beta Test"]}})
+          (play-from-hand state :corp "Gatekeeper" "HQ")
+          (take-credits state :corp)
+          (let [hand    (:hand (get-corp))
+                gate    (get-ice state :hq 0)
+                log-str (str "Corp uses Gatekeeper to shuffle"
+                             ".+"
+                             "from HQ into R&D")]
+              (run-on state "HQ")
+              (rez state :corp gate)
+              (run-continue state)
+              (card-subroutine state :corp gate 0)
+              (click-prompt state :corp "0")
+              (click-card state :corp (find-card "Accelerated Beta Test" hand))
+              (click-card state :corp (find-card "Project Vitruvius" hand))
+              (click-card state :corp (find-card "NEXT Wave 2" hand))
+              (is (last-log-contains? state log-str)))))
+      (testing "with cards chosen only from Archives"
+        (do-game
+          (new-game {:corp {:deck [(qty "Ice Wall" 10)]
+                            :hand ["Gatekeeper"]
+                            :discard ["Efficiency Committee"
+                                      "Hyperloop Extension"
+                                      "Bifrost Array"]}})
+          (play-from-hand state :corp "Gatekeeper" "HQ")
+          (take-credits state :corp)
+          (let [discard    (:discard (get-corp))
+                gate    (get-ice state :hq 0)
+                log-str (str "Corp uses Gatekeeper to shuffle"
+                             ".+"
+                             "from Archives into R&D")]
+              (run-on state "HQ")
+              (rez state :corp gate)
+              (run-continue state)
+              (card-subroutine state :corp gate 0)
+              (click-prompt state :corp "0")
+              (click-card state :corp (find-card "Efficiency Committee" discard))
+              (click-card state :corp (find-card "Hyperloop Extension" discard))
+              (click-card state :corp (find-card "Bifrost Array" discard))
+              (is (last-log-contains? state log-str)))))
+      (testing "with cards chosen both from HQ and R&D"
+        (do-game
+          (new-game {:corp {:deck [(qty "Ice Wall" 10)]
+                            :hand ["Gatekeeper"
+                                   "Project Vitruvius"
+                                   "NEXT Wave 2"
+                                   "Accelerated Beta Test"]
+                            :discard ["Efficiency Committee"
+                                      "Hyperloop Extension"]}})
+          (play-from-hand state :corp "Gatekeeper" "HQ")
+          (take-credits state :corp)
+          (let [hand    (:hand (get-corp))
+                discard (:discard (get-corp))
+                gate    (get-ice state :hq 0)
+                log-str (str "Corp uses Gatekeeper to shuffle"
```

Comment:
I understand the desire for clarity, but this could easily be one or two lines, not 7 lol.